torch_brain/utils/stitcher.py

```diff
@@ -71,7 +74,159 @@ def stitch(timestamps: torch.Tensor, values: torch.Tensor) -> torch.Tensor:
     return averages
 
 
-class StitchEvaluator(L.Callback):
+class DecodingStitchEvaluator(L.Callback):
+    r"""A convenient stitching and evaluation framework to use when:
+     1. Your model outputs have associated timestamps
+     2. And your sampling strategy involves overlapping time windows, requiring
+        stitching to coalesce the predictions and targets before computing the
+        evaluation metric.
+     3. (Optional) You are training on multiple sessions/recordings and want to
+        compute metrics for each session individually.
+
+    This callback handles the stitching of the predictions and targets for each
+    session and computes the metric for each session. The average metric value
+    across all sessions is also computed and logged.
+    Since the stitching is done only on tensors on the same GPU, sequences that are
+    split across multiple GPUs will not be stitched together. In this case, it is
+    recommended to use a stitching-aware sampler like
+    `DistributedStitchingFixedWindowSampler` which ensures that the sequences are
+    split across GPUs in a way that allows for correct stitching.
+
+    This callback is called _after_ the validation_step, and has two main inputs:
+    - The output of the validation_step function, which is expected to be the
+      model predictions. These should be a :class:`~torch.Tensor` of shape (B, N, ...),
+      where B is the batch size, N is the number of timestamps.
+    - The batch dictionary that should have atleast the following keys:
+        - "target_values": :class:`~torch.Tensor` of shape (B, N, ...)
+        - "output_timestamps": :class:`~torch.Tensor` of shape (B, N) and dtype float
+        - "output_mask": :class:`~torch.Tensor` of shape (B, N) and dtype bool
+        - "session_id": A list of session IDs for each sample in the batch
+        - "absolute_start": :class:`~torch.Tensor` of shape (B) and dtype float
+    Please refer to the examples/poyo/train.py script for an example of how to write
+    a validation_step(...) function that outputs the required tensors.
+
+    This callback operates by maintaining a cache of the predictions, targets, and
+    timestamps for each session. This cache is updated at the end of each batch.
+    Finally, at the end of the epoch, the cache is coalesced and the metric is
+    computed for each session.
+    """
+
+    def __init__(
+        self,
+        session_ids: Iterable[str],
+        modality_spec: Optional[ModalitySpec] = None,
+        metric_factory: Optional[Callable[[int], ModalitySpec]] = None,
```

Comment:
what's `metric_factory`? this nullifies the option to specify a different metric per dataset?

Reply:
are we creating a concept that will not work for POYOPlus?

Reply:
`metric_factory` is when you wanna use a custom metric bypassing the modality_spec framework.

Reply:
This will not work for POYOPlus. This stitcher is not meant for handling multi-task output

Reply:
I think I have made a mistake here. metric_factor's type should be `Optional[Callable[[int], torchmetrics.Metric]`

Reply:
this is a new concept. I would rather it be in a separate PR/discussion. For now, I do not see why are introducing two completely different ways for using the stitcher, based on whether it's POYO or POYOPlus. It will be extremely confusing.

Reply:
is your issue with the existence of `metric_factory` as an optional parameter?

Reply:
removed